File: backend/app/schemas/classroom.py
from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class RosterStudentIn(BaseModel):
    name: str = Field(max_length=100)
    surname: str = Field(max_length=100)
    school_id: int

    @field_validator("name", "surname")
    @classmethod
    def not_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("must not be blank")
        return value.strip()

    @field_validator("school_id")
    @classmethod
    def positive(cls, value: int) -> int:
        if value <= 0:
            raise ValueError("must be a positive number")
        return value


class ClassroomCreate(BaseModel):
    name: str = Field(max_length=100)
    students: list[RosterStudentIn] = []


class ClassroomOut(BaseModel):
    id: int
    name: str
    student_count: int
    created_at: datetime

    model_config = {"from_attributes": True}


class ClassroomBrief(BaseModel):
    id: int
    name: str

    model_config = {"from_attributes": True}


class RosterImportRequest(BaseModel):
    students: list[RosterStudentIn]


class PdfImportClassroomResult(BaseModel):
    classroom_id: int
    classroom_name: str
    created_classroom: bool
    students_added: int
    skipped: list[str] = []


class PdfImportResult(BaseModel):
    classrooms: list[PdfImportClassroomResult]
    errors: list[str] = []

File: frontend/lib/models/classroom.dart
class ClassroomOut {
  final int id;
  final String name;
  final int studentCount;
  final DateTime createdAt;

  ClassroomOut({
    required this.id,
    required this.name,
    required this.studentCount,
    required this.createdAt,
  });

  factory ClassroomOut.fromJson(Map<String, dynamic> json) {
    return ClassroomOut(
      id: json['id'] as int,
      name: json['name'] as String,
      studentCount: json['student_count'] as int,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class RosterStudentIn {
  final String name;
  final String surname;
  final int schoolId;

  RosterStudentIn({required this.name, required this.surname, required this.schoolId});

  Map<String, dynamic> toJson() => {
        'name': name,
        'surname': surname,
        'school_id': schoolId,
      };
}

class PdfImportClassroomResult {
  final int classroomId;
  final String classroomName;
  final bool createdClassroom;
  final int studentsAdded;
  final List<String> skipped;

  PdfImportClassroomResult({
    required this.classroomId,
    required this.classroomName,
    required this.createdClassroom,
    required this.studentsAdded,
    required this.skipped,
  });

  factory PdfImportClassroomResult.fromJson(Map<String, dynamic> json) {
    return PdfImportClassroomResult(
      classroomId: json['classroom_id'] as int,
      classroomName: json['classroom_name'] as String,
      createdClassroom: json['created_classroom'] as bool,
      studentsAdded: json['students_added'] as int,
      skipped: (json['skipped'] as List).map((e) => e as String).toList(),
    );
  }
}

class PdfImportResult {
  final List<PdfImportClassroomResult> classrooms;
  final List<String> errors;

  PdfImportResult({required this.classrooms, required this.errors});

  factory PdfImportResult.fromJson(Map<String, dynamic> json) {
    return PdfImportResult(
      classrooms: (json['classrooms'] as List)
          .map((e) => PdfImportClassroomResult.fromJson(e as Map<String, dynamic>))
          .toList(),
      errors: (json['errors'] as List).map((e) => e as String).toList(),
    );
  }
}
